---
title: "Top-hit identity distribution"
author: "Colin J. Brislawn"
date: "May 26, 2023"
output: html_document
---

```{r setup0, include=FALSE}
# library(phyloseq)
library("tidyverse")
# library("vegan")
library("kableExtra")

knitr::opts_chunk$set(echo = TRUE, cache = F)

theme_set(theme_bw())
theme_update(
  strip.background = element_blank(),
  axis.text.x = element_text(angle = -20, hjust = 0)
)
```

```{r lint, include=F, eval=F}
library(styler)
styler:::style_active_file()

library(lintr)
lintr::lint("THID.Rmd",
  linters = with_defaults(
    line_length_linter = line_length_linter(120),
    object_name_linter = NULL,
    commented_code_linter = NULL
  )
)
```

## Goal

Make the [Top-hit identity distribution plots](https://www.drive5.com/usearch/manual/thid.html) from Qiime2 vsearch hits.

[x-link to form post](https://forum.qiime2.org/t/understanding-naive-bayes-sklearn-classifier-top-hit-identity-distribution/26584)

```{bash, eval = FALSE}
cd /mnt/c/Users/cbris/git-repos/q2-forums/top-hit-identity

conda activate qiime2-2023.5

qiime feature-classifier vsearch-global \
  --i-query rep-seqs.qza \
  --i-reference-reads silva-138-99-seqs-515-806.qza \
  --p-top-hits-only \
  --p-maxhits 1 \
  --p-maxaccepts 20 \
  --p-maxrejects 'all' \
  --p-perc-identity 0.6 --p-query-cov 0.6 \
  --p-output-no-hits \
  --p-threads 4 \
  --o-search-results vsearch-blast6.qza

unzip -oj vsearch-blast6.qza -d vsearch-blast6-mvp


qiime feature-classifier vsearch-global \
  --i-query dada2_rep_set.qza \
  --i-reference-reads silva-138-99-seqs-515-806.qza \
  --p-top-hits-only \
  --p-maxhits 1 \
  --p-maxaccepts 20 \
  --p-maxrejects 'all' \
  --p-perc-identity 0.6 --p-query-cov 0.6 \
  --p-output-no-hits \
  --p-threads 4 \
  --o-search-results vsearch-blast6.qza

unzip -oj vsearch-blast6.qza -d vsearch-blast6-pdmice

rm -f vsearch-blast6.qza
```


```{r}
load_blast6 <- function(path) {
  read_tsv(path,
    # Col_names from https://www.metagenomics.wiki/tools/blast/blastn-output-format-6
    col_names = c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore")
  ) %>%
    # Add colors based on https://www.drive5.com/usearch/manual/thid.html
    # I don't beleave for a second that the species concept can be defined by a set % identity,
    # but that's really more of an issue with the species concept.
    mutate(color = case_when(
      pident == 100 ~ "Species",
      pident >= 98 ~ "Genus",
      pident >= 93 ~ "Family",
      pident >= 91 ~ "Order",
      pident >= 82 ~ "Class",
      pident >= 80 ~ "Phylum"
    )) %>%
    mutate(color = factor(color, levels = c("Phylum", "Class", "Order", "Family", "Genus", "Species"))) %>%
    return()
}

plot_thid <- function(df) {
  df %>%
    group_by(pident, color) %>%
    summarize(nFeatures = n()) %>%
    ggplot(aes(x = fct_rev(as_factor(pident)), y = nFeatures, fill = color)) +
    geom_bar(stat = "identity") +
    geom_text(aes(label = nFeatures), vjust = -0.5) +
    scale_y_sqrt() %>%
    return()
}

load_blast6("vsearch-blast6-mvp/blast6.tsv") %>%
  plot_thid() +
  labs(
    title = "Top-hit Identity Distribution",
    subtitle = "Moving Pictures DADA2 to silva-138-99-seqs-515-806",
    fill = "Robert Edgar's \ntaxa cutoffs"
  )

load_blast6("vsearch-blast6-pdmice/blast6.tsv") %>%
  plot_thid() +
  labs(
    title = "Top-hit Identity Distribution",
    subtitle = "PD-mice DADA2 to silva-138-99-seqs-515-806",
    fill = "Robert Edgar's \ntaxa cutoffs"
  )
```
